synapse/logging/opentracing.py

```diff
@@ -733,37 +733,59 @@ def decorator(func):
 
         _opname = opname if opname else func.__name__
 
-        @wraps(func)
-        def _trace_inner(*args, **kwargs):
-            if opentracing is None:
-                return func(*args, **kwargs)
+        if inspect.iscoroutinefunction(func):
+            @wraps(func)
+            async def _trace_inner(*args, **kwargs):
+                if opentracing is None:
+                    return await func(*args, **kwargs)
 
-            scope = start_active_span(_opname)
-            scope.__enter__()
+                scope = start_active_span(_opname)
+                scope.__enter__()
```

Comment:
You've made this harder than it needs to be :).

No need to call `__enter__`/`__exit__` manually. Just:

```
with scope:
    return await func(*args, **kwargs)
```

possibly with an `except` clause just to `set_tag(tags.ERROR)`

Reply:
Yeah, I was copying and pasting from below. I can update both to actually use a context manager instead.

Reply:
Ah, I see the `Deferred` case can't use the context manager easily, that's why it was that way. 👍

Reply:
We could always assume that the non-coroutine case was an inlineDeferred, this seems to be what `measure_func` does:

https://github.com/matrix-org/synapse/blob/39230d217104f3cd7aba9065dc478f935ce1e614/synapse/util/metrics.py#L64-L87